### PR TITLE
AP_Compass_AK8963:

### PR DESCRIPTION
**Pull request by dgrat**
Was reading how the new compass driver works and just tried to enhance the readability of the code by reducing the deepness of indentations.
I generally try to avoid arrow head like code elements by inverting the checks if possible. 
I think this helps to reduce maintenance of the code. 
May yield also slightly enhanced performance due to loss of complexity.

Signed-off-by: Daniel Frenzel <dgdanielf@gmail.com>

### CODE REVIEW

**Comment by staroselskii** [comment]
@dgrat Thanks! It does indeed make the code more readable. It won't break anything so I consider it ready to merge.

**Comment by lucasdemarchi** [comment]
@dgrat @staroselskii I had some pending changes to AK8963. I cherry-picked your patch (the parts that actually applied with the current state of the tree) and rebased my branch on it.  I'm going to send it shortly: https://github.com/lucasdemarchi/ardupilot/commits/pr-wip/ak8963

@dgrat Are you ok with your changed commit there?

**Comment by staroselskii** [comment]
@lucasdemarchi I'm working on the code as well right now. 

I took a slightly different approach, though. It'll be very difficult to merge them later. I suggest you take a look at https://github.com/staroselskii/ardupilot/commits/wip/ak8963-code-style-fix 

I find it easier to read when methods are small. Maybe cherry-pick my changes if you don't mind?

**Comment by lucasdemarchi** [comment]
> I find it easier to read when methods are small. Maybe cherry-pick my changes if you don't mind?

Long methods are hard to follow but in this specific cases the smaller ones are worse IMO. It leads to our current state in which we reset the AK8963 for each measurement, check the WAI in the wrong order, have no annotation on methods that lock or need the lock to be taken, etc.  While the first commit from @dgrat is to make it better style-wise, the last commit is fixing a real issue we have today. And in that one I preferred to move `_collect_samples()` to be inside the `update()` method rather than making the methods smaller because `update()` is the only (and needs to be the only) caller.

**Comment by staroselskii** [comment]
@lucasdemarchi 
> Long methods are hard to follow but in this specific cases the smaller ones are worse IMO. It leads to our current state in which we reset the AK8963 for each measurement, check the WAI in the wrong order, have no annotation on methods that lock or need the lock to be taken, etc. 

I guess it's a judgement call. As for me if a method does several things (i.e. filtering the raw value and adjusting it appropriately), it should be split. Since we're moving to SI units I believe all data manipulation should be pipelined. (it's purer and hence easier to reason about). The data flow is more obvious that way. But I guess it depends, doesn't it? 

There is a commit message I don't entirely understand (87db9af):
> Instead of sending the I2C over SPI commands every time we want to get a
sample from AK8963, just let the MPU9250 configured so it gets the
samples and put them in its internal registers. 

The MPU9250 is already configured that way. A new request is issued if and only if ```_collect_samples()``` fails. The state is not updated otherwise. That's why I don't really see the point of reimplementing the ```_update()``` method. 

Sorry if I'm being too serious about it. It's just a few lines. I also wanted to ask what ```WAI``` stands for. I guess I haven't come across this abbreviation.

**Comment by dgrat** [comment]
@lucasdemarchi @staroselskii 
I am okay with the changes. It was anyway just a 5 min style commit.

**Comment by lucasdemarchi** [comment]
> But I guess it depends, doesn't it?

It depends, yes... in some places I prefer having a long method to ensure nobody else will call that method.

> The MPU9250 is already configured that way. A new request is issued if and only if _collect_samples() fails. The state is not updated otherwise. That's why I don't really see the point of reimplementing the _update() method.

Wow... the message in that commit reflected my memory back when I wrote that code some time ago. I was cleaning it up on top of master.  It seems it's not true anymore that we are resetting the chip for each sample as we was.  I remember adding some printfs() on "_start_conversion()" and getting pretty much once for each sample. Maybe that was before 7a417d1 (AP_Compass: AK8963 rework)?

I still think doing it like a state machine is the wrong thing to do, though. It's not like BMP085 in which you trigger the conversion and the next time you read the values. Calling `_start_conversion()` should only happen once.  I'll change that commit accordingly.

And... we probably want to set I2C_MST_DLY. The datasheet here is confusing like saying it would sample the AK8963 at SAMPLE_RATE as per register 25.  This is really not physically possible since for us that would give 8kHz.

> Sorry if I'm being too serious about it. It's just a few lines. I also wanted to ask what WAI stands for. I guess I haven't come across this abbreviation.

Discussing the code is always the right thing to do ;-). WAI was a typo, should be "WIA" as per the datasheet, the device id. My point is that we were setting

**Comment by staroselskii** [comment]
> It depends, yes... in some places I prefer having a long method to ensure nobody else will call that method.

I don't see a problem calling```_make_factory_sensitivity_adjustment()``` multiple times. It doesn't change state. I guess I should've marked it as ```const```. I'm being stubborn but I really want to have this code as readable as possible before making the compass unit transition. (The HMC5883 driver is far more daunting from this perspective, though. And it also needs to be revisited. ).

> Maybe that was before 7a417d1 (AP_Compass: AK8963 rework)?

Yeah, I guess it was fixed in this one.

> I still think doing it like a state machine is the wrong thing to do, though.

I've thought that an ```update()``` method makes a perfect pair with a state machine. If something's wrong, we enter the error state. This way we can recover in a clear way. 

> And... we probably want to set I2C_MST_DLY. The datasheet here is confusing like saying it would sample the AK8963 at SAMPLE_RATE as per register 25. This is really not physically possible since for us that would give 8kHz. 

I have a setup which allows to monitor I2C transactions between the MPU9250 and the AK8963. I can check the actual speed if that's of any interest.

> It leads to our current state in which we reset the AK8963 for each measurement, check the WAI in the wrong order, have no annotation on methods that lock or need the lock to be taken, etc.

Do we check WIA in the wrong order? Either way, I wanted to ask what's the preferred and clearer approach of dealing with methods that need a lock to be taken? This is why you want to have a single method that acquires the lock, don't you? This way it's less probable to hit a deadlock, right?

**Comment by staroselskii** [comment]
@dgrat 

We are all in the same boat! I guess I had copy-pasted the code which deals with the lock from somewhere around APM codebase and never got around to fix it. Thanks! My approval on this one is not a final one, though. Because the code is shared between all the vehicles and boards.

**Comment by lucasdemarchi** [comment]
>> It depends, yes... in some places I prefer having a long method to ensure nobody else will call that method.
> I don't see a problem calling_make_factory_sensitivity_adjustment() multiple times. It doesn't change state. I guess I should've marked it as const. I'm being stubborn but I really want to have this code as readable as possible before making the compass unit transition.

That one is actually good. Right now I'm looking at your branch and seeing I can get that patch on top of mine.

>> I still think doing it like a state machine is the wrong thing to do, though.
> I've thought that an update() method makes a perfect pair with a state machine. If something's wrong, we enter the error state. This way we can recover in a clear way.

If it returns false, as per the datasheet you don't reset the chip or ask for a new measurement. You just discard that measurement.  And even if you had to do anything, why delaying for the next call to update()?  That's what the state machine was doing there.  We could very well just do something if the function returned false.

>> And... we probably want to set I2C_MST_DLY. The datasheet here is confusing like saying it would sample the AK8963 at SAMPLE_RATE as per register 25. This is really not physically possible since for us that would give 8kHz.
> I have a setup which allows to monitor I2C transactions between the MPU9250 and the AK8963. I can check the actual speed if that's of any interest.

The SCL should be giving 400 kHz, which is actually what we are setting.  However I'm interested to know how frequently we try to read something from the AK8963. Can you monitor the AUX_SDA line for the time between any 2 reads?

>> It leads to our current state in which we reset the AK8963 for each measurement, check the WAI in the wrong order, have no annotation on methods that lock or need the lock to be taken, etc.
> Do we check WIA in the wrong order?

yes... I'm just splitting the commits better with smaller fixes.

> Either way, I wanted to ask what's the preferred and clearer approach of dealing with methods that need a lock to be taken? This is why you want to have a single method that acquires the lock, don't you? This way it's less probable to hit a deadlock, right?

This is one of the reasons. A deadlock is actually easy to get because things just stop to work and we can't even start if we have one. Races (and we do have one on AK8963) and  starvations are more difficult. Leaving these parts on a single method make it easier to verify what's actually happening.

**Comment by lucasdemarchi** [comment]
Closing as agreed. See #2595.
